### PR TITLE
[BACK-3030] Persist latest time zone offset for Dexcom Connection into data set

### PR DESCRIPTION
**Pull request by darinkrauss**
- Determine time zone offset for Dexcom Connection from most recent datum
- Persist time zone offset for Dexcom Connection in data set
- Use display time zone offset, if available, in time zone calculation
- https://tidepool.atlassian.net/browse/BACK-3030

Note: I attempted to add tests for the key piece of this change, but ran into challenges. First, there are currently no tests at all for this code (ultimately, my bad). Second, I now have a whole set of tests that incorporate this change, as well as significant other changes (for related Dexcom work), but changing those tests to allow them to work just for this one change (without the rest) would be a significant lift and effectively throw-away (since it would be soon replaced by the new tests).

### CODE REVIEW

File: data/types/base_test.go
Comment (jh-bate) [inline data/types/base_test.go:976]: there is never a chance we are always just comparing nil is nil as its never set?

**Review by jh-bate — APPROVED** [review]
one tiny question and you have already answered the other "where the tests at" question :)

File: dexcom/fetch/runner.go
Comment (toddkazakov) [inline dexcom/fetch/runner.go:730]: Should there be a `break` here?

File: data/types/base_test.go
Comment (darinkrauss) [reply data/types/base_test.go:976]: Yes, that is possible. This test is just to check that the function `GetTimeZoneOffset` returns the same value (`nil` or otherwise) as stored in the `TimeZoneOffset` field. The data for this specific test, though, is non-`nil`, specifically set in call to `dataTypesTest.RandomBase()`.

File: dexcom/fetch/runner.go
Comment (darinkrauss) [reply dexcom/fetch/runner.go:730]: The `for` loop has a `&& timezoneOffset == nil` condition to exit the loop once the first non-`nil` `timezoneOffset` is found. 

However, given your comment, it is non-obvious, therefore I'll update to make it more readable.

Comment (darinkrauss) [reply dexcom/fetch/runner.go:732]: Updated to make it more clear.

**Comment by darinkrauss** [comment]
This adds considerably more work in order to parse Dexcom date/time strings correctly and, more importantly, to capture whether the original date/time string included a time zone specifier or not. While the attempt to parse the date/time string is a tad over-lenient, given what I've seen with real production Dexcom data, hopefully this will prevent any date/time parsing issues (and parse them as correctly as possible).

If the time zone was specified on the display time, then the time zone offset is forced to that with an appropriate change to the conversion offset (so the UTC bootstrapping calculations remain correct). This gives us the best change of have the time zone offset correctly represent 
"where the user is". It also fixes a current bug where all data from iOS (likely Android, too) has a zero time zone offset.

File: dexcom/fetch/translate.go
Comment (ewollesen) [inline dexcom/fetch/translate.go:47]: **Not asking for a change here, just hoping you can clarify something for me.**

Is this being exported just so you can call it directly from your test package?

Are you a fan of the separate test package paradigm? I've never understood it. Could you explain the benefits as you see them to me?

*If you said "no" to my first question of the previous paragraph, then you can probably just mark this resolved now and move on.* :)

I'll try to explain my current understanding of the concept. If you have a different understanding, I would be glad to learn from your understanding on the matter.

I understand that the point of having a separately named test package is to enforce that one can only test things via a minimal exported interface. There are two goals of this restriction: 

1. Maintain a minimal exported/public interface.
2. Maintain the ability to refactor the implementation of said interface without breaking tests. 

Does that align with your understanding?

I don't think that works in Go, does it? Go packages make a lot of use of dependency injection, and many of the injected objects are test mocks that specifically check implementation. Take for example the popular [gomock](https://pkg.go.dev/github.com/golang/mock/gomock) package[^1].

If the aim of having a separate package aim is to make it harder to test an interface's implementation (goal 2 above), then it seems like it's failed at that goal; dependency injection is widespread in Go, and there are numerous examples of mock DI objects being used to test implementation.

Moving on to the second goal, as soon as a single function is exported that isn't part of that minimal exported interface, goal 1 above is broken[^2]. That said, there are plenty of cases in our code were function are exported only so they can be tested.

At this point, neither of the stated goals are reached, yet there's significant friction around testing that results from the extra test package. It feels like there's no upside.

So to summarize, I feel like the goals of testing behavior and not implementation, and maintaining a minimal exported interface are good ones, but that using a separate test package doesn't bring a project closer to those goals, in my experience, it likely just makes it harder. For now the best solution might be to aim for disciplined test writing and code reviews to identify tests that are brittle due to testing implementation, and to rewrite them in better ways. 

[1]: I think Gomock is a useful tool, but I believe it's often misued in cases like this.

[2]: This seems to happen a lot. I assume because it's often hard to test an internal function via an external interface. The extra setup and teardown of internal state is a pain. Maybe that's supposed to be an indicator that the internal function should be in its own package? But that doesn't feel very Go like at all to me.

File: dexcom/fetch/runner.go
Comment (darinkrauss) [reply dexcom/fetch/runner.go:721]: I don't see how that would happen given these lines 712-714 above. Specifically:
```
if endIndex > length {
    endIndex = length
}
```
What am I missing?

File: dexcom/fetch/translate.go
Comment (darinkrauss) [reply dexcom/fetch/translate.go:60]: Yes, if there wasn't a Payload before, create one here. I don't think Payload will currently exist prior to this translate function being called, but just a bit of defensive coding here. Thus, if someone unfamiliar with this code sets up a Payload prior to calling this translate function, this code will not stomp on that pre-existing Payload.

File: dexcom/fetch/translate.go
Comment (darinkrauss) [reply dexcom/fetch/translate.go:47]: Not sure I agree with your interpretation. It is in the language specification and thus a feature to be used. What it really means is "modify the thing in place" which is what I want to do here. I don't want to create a copy (and all that entails - deep or shallow copy, ...).

Comment (ewollesen) [reply dexcom/fetch/translate.go:47]: *To be clear, not asking for anything to change, just wanting to understand if I'm missing some perspective on how using test packages are in any way useful.*

> I do tend to favor separate test packages so internal implementation details aren't generally tested.

Am I missing something, or does exporting a function solely so it can be tested, exactly contradict your stated reason for having a test package? As I see it a non-exported function is, by definition, "internal implementation details".

> From my perspective, a pragmatic accommodation to balance competing needs.

I'm all for pragmatism. I understand the pain of all that work to test things sometimes. 

What I don't understand, is what need is fulfilled with the test package? 

It can't be the prevention of testing implementation details, as that's not enforced by the test package, and even if it were, it's side-stepped regularly.

> At this point, as platform only uses the separate test package paradigm, I don't think changing it either globally or even for just this package makes sense.

Understood, not trying to ask for any changes, just trying to see if there's some benefit to the practice that I haven't understood.

Comment (ewollesen) [reply dexcom/fetch/translate.go:47]: > It is in the language specification and thus a feature to be used. 

I didn't say it wasn't, in fact I pointed out a place where it is used. Certainly you can understand that just because a feature exists, doesn't mean it should be used. Modifying passed arguments is certainly possible, but it's relatively rare, and especially given the name of the function, I don't think it's at all expected.

> What it really means is "modify the thing in place" which is what I want to do here.

Yes, so, since you don't want to make a copy/clone (which is totally fine), would you be opposed to a name that makes it obvious that `datum` is being modified, something like UpdateDatumTimes perhaps?

I wonder if this function wouldn't make more sense as a method on dataTypes.Base, a call like `datum.UpdateFromTimes(systemTime *dexcom.Time, displayTime *dexcom.Time)` is pretty clear in what it does.

File: dexcom/fetch/runner.go
Comment (ewollesen) [reply dexcom/fetch/runner.go:721]: Nothing. GitHub declined to show me that part, and when I clicked expand, I somehow missed those lines. My apologies.

File: dexcom/fetch/translate.go
Comment (darinkrauss) [reply dexcom/fetch/translate.go:47]: I suppose the root "problem" here is that the translation code is in the same package as the runner. If I had simply created a "dexcom/fetch/translate" package and exported all of the various translation functions from that, then we wouldn't be having this discussion. Yes, I suppose I could do that after-the-fact, but that seems like more work than benefit.

Comment (darinkrauss) [reply dexcom/fetch/translate.go:47]: I suppose I could change the name if that will clear this up for you. Adding the function to the `dataTypes.Base` type would be mixing in the `dexcom` package into a whole pile of code where it absolutely does not belong. (If Golang had a way to locally extend a struct, then maybe, but it does not.)

File: dexcom/fetch/runner.go
Comment (darinkrauss) [reply dexcom/fetch/runner.go:721]: No worries, just wanted to make sure I wasn't missing anything. Thanks!

File: dexcom/fetch/translate.go
Comment (jh-bate) [reply dexcom/fetch/translate.go:47]: > Adding the function to the dataTypes.Base type would be mixing in the dexcom package ...

This would be my sentiment with regards to `datum.UpdateFromTimes(`. 

It might not be perfect but we are using this pattern  see with the `func TranslateTime(` often already so its nothing different

**Review by jh-bate — DISMISSED** [review]
One query re a param name but LGTM

Well done on untangling as best as possible the mess that is dexcom dates!

**Review by toddkazakov — DISMISSED** [review]

File: dexcom/fetch/translate.go
Comment (ewollesen) [reply dexcom/fetch/translate.go:47]: Yeah, re-reading this today, I realize I was being a doofus, sorry.

Comment (ewollesen) [reply dexcom/fetch/translate.go:47]: Thanks for taking the time to answer my questions, I appreciate it.

File: dexcom/time.go
Comment (darinkrauss) [reply dexcom/time.go:202]: Ha! :) No, that would be a made up substitution. I didn't want to use `default` because it is a keyword, so this is just an alternate spelling that sounds about the same. Just like `bites` (for `bytes`, which is package), `lgr` (for `logger`, which is a package), etc...

File: dexcom/time.go
Comment (darinkrauss) [reply dexcom/time.go:55]: :+1: I like your approach much better so I merged it as-is into this PR. Thanks!

**Comment by darinkrauss** [comment]
@toddkazakov @jh-bate Just need a quick approval after I merged in Todd's alternate implementation of time parsing.

**Review by toddkazakov — APPROVED** [review]

**Review by ewollesen — APPROVED** [review]
LGTM

**Pull request merged**
